src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs

```diff
@@ -156,10 +156,15 @@ private bool Visit(ModelMetadata metadata, string key, object model)
         {
             RuntimeHelpers.EnsureSufficientExecutionStack();
 
-            if (model != null && !_currentPath.Add(model))
+            if (model != null)
             {
-                // This is a cycle, bail.
-                return true;
+                if (_currentPath.Contains(model))
+                {
+                    // This is a cycle, bail.
+                    return true;
+                }
+
+                _currentPath.Push(model);
```

Comment:
Why not keep the current pattern? This is going to hash the object twice if you're using the hashset.